go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,242 @@
+	if err != nil {
+		klog.Warningf("Unable to allocate subnet's management IP: %s", mgmtIfAddr.IP)
+		return err
+	}
+	if config.HybridOverlay.Enabled {
+		hybridOverlayIfAddr := util.GetNodeHybridOverlayIfAddr(subnet)
+
+		err = ipam.Allocate(hybridOverlayIfAddr.IP)
+		if err != nil {
+			klog.Warningf("Unable to allocate subnet's hybrid overlay interface IP: %s", hybridOverlayIfAddr.IP)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Initializes a new logical switch manager
+func newLogicalSwitchManager() *logicalSwitchManager {
+	return &logicalSwitchManager{
+		cache:    make(map[string]logicalSwitchInfo),
+		RWMutex:  sync.RWMutex{},
+		ipamFunc: NewIPAMAllocator,
+	}
+}
+
+// AddNode adds/updates a node to the logical switch manager for subnet
+// and IPAM management.
+func (manager *logicalSwitchManager) AddNode(nodeName string, hostSubnets []*net.IPNet) error {
+	manager.Lock()
+	defer manager.Unlock()
+	if lsi, ok := manager.cache[nodeName]; ok && !reflect.DeepEqual(lsi.hostSubnets, hostSubnets) {
+		klog.Warningf("Node %q logical switch already in cache with subnet %s; replacing with %s", nodeName,
+			util.JoinIPNets(lsi.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	}
+	var ipams []ipam.Interface
+	for _, subnet := range hostSubnets {
+		ipam, err := manager.ipamFunc(subnet)
+		if err != nil {
+			klog.Warningf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			return err
+		}
+		ipams = append(ipams, ipam)
+	}
+	manager.cache[nodeName] = logicalSwitchInfo{
+		hostSubnets:  hostSubnets,
+		ipams:        ipams,
+		noHostSubnet: len(hostSubnets) == 0,
+	}
+
+	return nil
+}
+
+// Remove a switch/node from the the logical switch manager
+func (manager *logicalSwitchManager) DeleteNode(nodeName string) {
+	manager.Lock()
+	defer manager.Unlock()
+	delete(manager.cache, nodeName)
+}
+
+// Given a switch name, checks if the switch is a noHostSubnet switch
+func (manager *logicalSwitchManager) IsNonHostSubnetSwitch(nodeName string) bool {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	return ok && lsi.noHostSubnet
+}
+
+// Given a switch name, get all its host-subnets
+func (manager *logicalSwitchManager) GetSwitchSubnets(nodeName string) []*net.IPNet {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	// make a deep-copy of the underlying slice and return so that there is no
+	// resource contention
+	if ok && len(lsi.hostSubnets) > 0 {
+		var subnets []*net.IPNet
+		for _, hsn := range lsi.hostSubnets {
+			subnet := *hsn
+			subnets = append(subnets, &subnet)
+		}
+		return subnets
+	}
+	return nil
+}
+
+// AllocateIPs will block off IPs in the ipnets slice as already allocated
+// for a given switch
+func (manager *logicalSwitchManager) AllocateIPs(nodeName string, ipnets []*net.IPNet) error {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	if ok && ipnets != nil && len(lsi.ipams) > 0 {
```

Comment:
just inverse this logic and return nil if true so you can move the for loops out of another level of nested blocks

Reply:
should we also print a log here debug, warning or otherwise if we are unable to AllocateIPs? Maybe even return an error if there are no ipams?

Reply:
it just reads weird, but sure: ```
if ipnets == nil || !ok || (ok && len(lsi.ipams) <=0) {
return nil
}
```

Reply:
perhaps a good idea to log and return error. will make the fix.